cmd/agent/app/agent.go

```diff
@@ -295,57 +298,53 @@ func startServiceExportController(ctx controllerscontext.Context) (bool, error)
-		return err
-	}
-
-	impersonatorSecret := &corev1.Secret{
-		ObjectMeta: metav1.ObjectMeta{
-			Namespace: opts.ClusterNamespace,
-			Name:      names.GenerateImpersonationSecretName(opts.ClusterName),
-		},
-	}
-
-	clusterObj, err := generateClusterInControllerPlane(controlPlaneRestConfig, opts, *impersonatorSecret)
-	if err != nil {
-		return err
-	}
-
-	clusterImpersonatorSecret, err := obtainCredentialsFromMemberCluster(clusterRestConfig, opts)
-	if err != nil {
-		return err
-	}
-
-	impersonatorSecret.OwnerReferences = []metav1.OwnerReference{
-		*metav1.NewControllerRef(clusterObj, clusterv1alpha1.SchemeGroupVersion.WithKind("Cluster"))}
-	impersonatorSecret.Data = map[string][]byte{
-		clusterv1alpha1.SecretTokenKey: clusterImpersonatorSecret.Data[clusterv1alpha1.SecretTokenKey]}
-	// create secret to store impersonation info in control plane
-	_, err = util.CreateSecret(controlPlaneKubeClient, impersonatorSecret)
-	if err != nil {
-		return fmt.Errorf("failed to create impersonator secret in control plane. error: %v", err)
-	}
-	return nil
-}
-
-func generateClusterInControllerPlane(controlPlaneRestConfig *restclient.Config, opts *options.Options, impersonatorSecret corev1.Secret) (*clusterv1alpha1.Cluster, error) {
+func generateClusterInControllerPlane(controlPlaneRestConfig, clusterConfig *restclient.Config, opts util.RegistryClusterOption, secret corev1.Secret, impersonatorSecret corev1.Secret) (*clusterv1alpha1.Cluster, error) {
 	clusterObj := &clusterv1alpha1.Cluster{ObjectMeta: metav1.ObjectMeta{Name: opts.ClusterName}}
-	mutateFunc := func(cluster *clusterv1alpha1.Cluster) {
+	var mutateFunc, tempFunc func(cluster *clusterv1alpha1.Cluster)
+	tempFunc = func(cluster *clusterv1alpha1.Cluster) {
 		cluster.Spec.SyncMode = clusterv1alpha1.Pull
 		cluster.Spec.APIEndpoint = opts.ClusterAPIEndpoint
 		cluster.Spec.ProxyURL = opts.ProxyServerAddress
 		cluster.Spec.ImpersonatorSecretRef = &clusterv1alpha1.LocalSecretReference{
 			Namespace: impersonatorSecret.Namespace,
 			Name:      impersonatorSecret.Name,
 		}
-	}
+		if opts.ClusterProvider != "" {
+			cluster.Spec.Provider = opts.ClusterProvider
+		}
+
+		if opts.ClusterZone != "" {
+			cluster.Spec.Zone = opts.ClusterZone
+		}
+
+		if opts.ClusterRegion != "" {
+			cluster.Spec.Region = opts.ClusterRegion
+		}
 
+		if clusterConfig.TLSClientConfig.Insecure {
+			cluster.Spec.InsecureSkipTLSVerification = true
+		}
+		if clusterConfig.Proxy != nil {
+			url, err := clusterConfig.Proxy(nil)
+			if err != nil {
+				klog.Errorf("clusterConfig.Proxy error, %v", err)
+			} else {
+				cluster.Spec.ProxyURL = url.String()
+			}
+		}
+	}
+	if opts.IsKubeCredentialsEnabled() {
+		mutateFunc = func(cluster *clusterv1alpha1.Cluster) {
+			tempFunc(cluster)
+			cluster.Spec.SecretRef = &clusterv1alpha1.LocalSecretReference{
+				Namespace: secret.Namespace,
+				Name:      secret.Name,
+			}
+		}
+	}
+	if opts.IsKubeImpersonatorEnabled() {
```

Comment:
What if opts.IsKubeImpersonatorEnabled() if false?

Reply:
if `IsKubeCredentialsEnabled` and `IsKubeImpersonatorEnabled` is false, mutateFunc is nil,  and return err

Reply:
get it！

Reply:
The logic here still seems to be a bit problematic. When `IsKubeImpersonatorEnabled` is false, mutateFunc will be nil.

Reply:
> When IsKubeCredentialsEnabled and IsKubeImpersonatorEnabled are false, we need to create a Cluster object without them.

mutateFunc is nil, but still can create cluster object successfully
`
func CreateOrUpdateClusterObject(controlPlaneClient karmadaclientset.Interface, clusterObj *clusterv1alpha1.Cluster, mutate func(*clusterv1alpha1.Cluster)) (*clusterv1alpha1.Cluster, error) {
	cluster, exist, err := GetClusterWithKarmadaClient(controlPlaneClient, clusterObj.Name)
	if err != nil {
		return nil, err
	}
	if mutate == nil {
		klog.Warningf("mutate is nil, report-secrets may be contains KubeCredentials or KubeImpersonator at least, otherwise create cluster object without secret field")
		if !exist {
			if cluster, err = createCluster(controlPlaneClient, clusterObj); err != nil {
				klog.Warningf("failed to create cluster(%s). error: %v", clusterObj.Name, err)
				return nil, err
			}
		}
	}
`

Reply:
Sorry, I don't understand clearly. When mutate is nill, do we just create an empty Cluster object?

Reply:
> When mutate is nill, do we just create an empty Cluster object?

Oh, that's really a problem

Reply:
review again pls？ @XiShanYongYe-Chang

Reply:
okay~

Reply:
When `report-secrets` is `{"KubeCredentials", "KubeImpersonator"}`, only impersonator secret will be report.

I suggest to update like this:

```go
--- a/cmd/agent/app/agent.go
+++ b/cmd/agent/app/agent.go
@@ -312,8 +312,7 @@ func startServiceExportController(ctx controllerscontext.Context) (bool, error)

 func generateClusterInControllerPlane(opts util.ClusterRegisterOption) (*clusterv1alpha1.Cluster, error) {
        clusterObj := &clusterv1alpha1.Cluster{ObjectMeta: metav1.ObjectMeta{Name: opts.ClusterName}}
-       var mutateFunc, tempFunc func(cluster *clusterv1alpha1.Cluster)
-       tempFunc = func(cluster *clusterv1alpha1.Cluster) {
+       mutateFunc := func(cluster *clusterv1alpha1.Cluster) {
                cluster.Spec.SyncMode = clusterv1alpha1.Pull
                cluster.Spec.APIEndpoint = opts.ClusterAPIEndpoint
                cluster.Spec.ProxyURL = opts.ProxyServerAddress
@@ -340,28 +339,22 @@ func generateClusterInControllerPlane(opts util.ClusterRegisterOption) (*cluster
                                cluster.Spec.ProxyURL = url.String()
                        }
                }
-       }
-       if opts.IsKubeCredentialsEnabled() {
-               mutateFunc = func(cluster *clusterv1alpha1.Cluster) {
-                       tempFunc(cluster)
+
+               if opts.IsKubeCredentialsEnabled() {
                        cluster.Spec.SecretRef = &clusterv1alpha1.LocalSecretReference{
                                Namespace: opts.Secret.Namespace,
                                Name:      opts.Secret.Name,
                        }
                }
-       }
-       if opts.IsKubeImpersonatorEnabled() {
-               mutateFunc = func(cluster *clusterv1alpha1.Cluster) {
-                       tempFunc(cluster)
+
+               if opts.IsKubeImpersonatorEnabled() {
                        cluster.Spec.ImpersonatorSecretRef = &clusterv1alpha1.LocalSecretReference{
                                Namespace: opts.ImpersonatorSecret.Namespace,
                                Name:      opts.ImpersonatorSecret.Name,
                        }
                }
        }
-       if mutateFunc == nil {
-               mutateFunc = tempFunc
-       }
```